gohelper.go

```diff
@@ -11,89 +11,347 @@ import (
 	Query    struct {
 		IsMaster bool `json:"is_master"`
 	}
 }
-type Relationships map[string][]Relationship
-
-type PlatformInfo struct {
-	Relationships Relationships
-	//Application     ApplicationInfo
-	//Routes          RouteInfo
-	//Variables       map[string]string
-	ApplicationName string
-	DocRoot         string
-	Branch          string
-	TreeId          string
-	AppDir          string
-	Environment     string
-	Project         string
-	Entropy         string
-	Socket          string
-	Port            string
-}
-
-// NewPlatformInfo returns a struct containing environment information
-// for the current Platform.sh environment. That includes the port on
-// which to listen for web requests, database credentials, and so on.
-// If that information is not available due to being called when not
-// running on Platform.sh an error will be returned.
-func NewPlatformInfo() (*PlatformInfo, error) {
-	p := &PlatformInfo{}
+
+type Credentials map[string][]Credential
+
+type Route struct {
+	OriginalUrl    string            `json:"original_url"`
+	Attributes     map[string]string `json:"attributes"`
+	Type           string            `json:"type"`
+	RestrictRobots bool              `json:"restrict_robots"`
+	Tls            struct {
+		ClientAuthentication         string   `json:"client_authentication"`
+		MinVersion                   int      `json:"min_version"`
+		ClientCertificateAuthorities []string `json:"client_certificate_authorities"`
+		StrictTransportSecurity      struct {
+			IncludeSubdomains bool `json:"include_subdomains"`
+			Enabled           bool `json:"enabled"`
+			Preload           bool `json:"preload"`
+		}
+	}
+	Upstream string `json:"upstream"`
+	Cache    struct {
+		Enabled    bool     `json:"enabled"`
+		Headers    []string `json:"headers"`
+		Cookies    []string `json:"cookies"`
+		DefaultTtl int      `json:"default_ttl"`
+	}
+	HttpAccess struct {
+		Addresses []string          `json:"addresses"`
+		BasicAuth map[string]string `json:"basic_auth"`
+	}
+	Primary bool   `json:"primary"`
+	Id      string `json:"id"`
+	Ssi     struct {
+		Enabled bool `json:"enabled"`
+	}
+
+	// This field is not part of the JSON definition, but it gets added
+	// to the struct from the JSON array key.
+	Url string
+}
+
+type Routes map[string]Route
+
+type PlatformConfig struct {
+	// Prefixed simple values, build or deploy.
+	applicationName string
+	treeId          string
+	appDir          string
+	project         string
+	entropy         string
+
+	// Prefixed simple values, runtime only.
+	branch       string
+	environment  string
+	documentRoot string
+	smtpHost     string
+	mode         string
+
+	// Prefixed complex values.
+	credentials Credentials
+	variables   envList
+	routes      Routes
+	application map[string]interface{}
+
+	// Unprefixed simple values.
+	socket string
+	port   string
+
+	// Internal data.
+	prefix string
+}
+
+func NewConfigReal(getter envReader, prefix string) (*PlatformConfig, error) {
```

Comment:
Don't export that if it shouldn't be part of the public API

Reply:
I'm torn on this one.  If I want to have the tests in a separate namespace, this will have to be public.  The tests are not in a separate namespace currently, mainly because of Go's funky namespace design I'm not sure how to reference a sibling namespace relatively without going through a full github.com path, which won't work until this branch is merged.  If I can figure that out, it's IMO a better test approach.  (The zeitgeist I found Googling would agree.)